File: transcribers/google.js
import fs from 'fs';

export class GoogleSTTTranscriber {
  constructor() {
    this.name = 'Google Speech-to-Text';
    this.language = 'lv'; // Latvian
    
    // For Vercel deployment, we'll need to use Google Cloud Speech-to-Text API
    // This requires a service account key and proper setup
    if (!process.env.GOOGLE_APPLICATION_CREDENTIALS || !process.env.GOOGLE_CLOUD_PROJECT_ID) {
      throw new Error('Google Cloud credentials not configured');
    }
  }

  async transcribe(audioFilePath) {
    try {
      // For now, return a placeholder since Google Cloud setup is complex
      // In production, you would:
      // 1. Upload audio to Google Cloud Storage or send directly
      // 2. Use the Speech-to-Text API with Latvian language code
      // 3. Return the transcript
      
      console.log('Google STT transcription requested for:', audioFilePath);
      
      // Simulate processing time
      await new Promise(resolve => setTimeout(resolve, 1000));
      
      return 'Google Speech-to-Text transcription requires Google Cloud setup. Please configure GOOGLE_APPLICATION_CREDENTIALS and GOOGLE_CLOUD_PROJECT_ID environment variables.';
      
    } catch (error) {
      throw new Error(`Google STT transcription failed: ${error.message}`);
    }
  }
}

// Alternative implementation using Google Cloud Speech-to-Text API
export class GoogleCloudSTTTranscriber {
  constructor() {
    this.name = 'Google Speech-to-Text';
    this.language = 'lv';
    
    if (!process.env.GOOGLE_CLOUD_API_KEY) {
      throw new Error('Google Cloud API key not configured');
    }
  }

  async transcribe(audioFilePath) {
    try {
      // Read the audio file
      const audioBuffer = fs.readFileSync(audioFilePath);
      const audioBase64 = audioBuffer.toString('base64');
      
      // Google Cloud Speech-to-Text API request
      const response = await fetch('https://speech.googleapis.com/v1/speech:recognize', {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${process.env.GOOGLE_CLOUD_API_KEY}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          config: {
            encoding: 'WEBM_OPUS', // Adjust based on your audio format
            sampleRateHertz: 48000,
            languageCode: 'lv', // Latvian
            enableAutomaticPunctuation: true,
            model: 'latest_long', // Use latest model
          },
          audio: {
            content: audioBase64,
          },
        }),
      });

      if (!response.ok) {
        throw new Error(`Google API error: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();
      
      if (data.results && data.results.length > 0) {
        return data.results[0].alternatives[0].transcript;
      } else {
        return 'No speech detected in audio file';
      }
      
    } catch (error) {
      throw new Error(`Google STT transcription failed: ${error.message}`);
    }
  }
}

File: transcribers/google.py
"""Google Speech-to-Text transcription service."""

import os
from pathlib import Path
from typing import Optional

try:
    from google.cloud import speech
    GOOGLE_AVAILABLE = True
except ImportError:
    GOOGLE_AVAILABLE = False

from .base import BaseTranscriber, TranscriptionError
from config import MODELS


class GoogleTranscriber(BaseTranscriber):
    """Google Speech-to-Text transcription service."""
    
    def __init__(self, credentials_path: Optional[str] = None, project_id: Optional[str] = None):
        super().__init__(
            name=MODELS["google"]["name"],
            language=MODELS["google"]["language"]
        )
        
        if not GOOGLE_AVAILABLE:
            raise ImportError("Google Cloud Speech library not installed. Install with: pip install google-cloud-speech")
        
        # Set up credentials
        credentials_path = credentials_path or MODELS["google"]["api_key"]
        project_id = project_id or os.getenv("GOOGLE_CLOUD_PROJECT")
        
        if credentials_path and os.path.exists(credentials_path):
            os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = credentials_path
        
        if not project_id:
            raise ValueError("Google Cloud project ID is required")
        
        try:
            self.client = speech.SpeechClient()
        except Exception as e:
            raise ValueError(f"Failed to initialize Google Speech client: {str(e)}")
    
    def transcribe(self, audio_file_path: Path) -> str:
        """Transcribe audio using Google Speech-to-Text."""
        if not self.validate_audio_file(audio_file_path):
            raise TranscriptionError(f"Invalid audio file: {audio_file_path}")
        
        try:
            # Read audio file
            with open(audio_file_path, "rb") as audio_file:
                content = audio_file.read()
            
            # Configure recognition
            audio = speech.RecognitionAudio(content=content)
            config = speech.RecognitionConfig(
                encoding=speech.RecognitionConfig.AudioEncoding.ENCODING_UNSPECIFIED,
                sample_rate_hertz=16000,  # Common sample rate
                language_code=self.language,
                enable_automatic_punctuation=False,  # Raw output as requested
                model="latest_long",  # Better for longer audio
                use_enhanced=True  # Use enhanced model if available
            )
            
            # Perform transcription
            response = self.client.recognize(config=config, audio=audio)
            
            # Extract text from results
            if response.results:
                # Combine all results
                transcript_parts = []
                for result in response.results:
                    if result.alternatives:
                        transcript_parts.append(result.alternatives[0].transcript)
                
                return " ".join(transcript_parts).strip()
            else:
                raise TranscriptionError("No transcription results returned")
                
        except Exception as e:
            self.logger.error(f"Google Speech-to-Text transcription failed: {str(e)}")
            raise TranscriptionError(f"Google Speech-to-Text transcription failed: {str(e)}")
    
    def _detect_audio_encoding(self, audio_file_path: Path) -> speech.RecognitionConfig.AudioEncoding:
        """Detect audio encoding from file extension."""
        extension = audio_file_path.suffix.lower()
        
        encoding_map = {
            ".wav": speech.RecognitionConfig.AudioEncoding.LINEAR16,
            ".flac": speech.RecognitionConfig.AudioEncoding.FLAC,
            ".mp3": speech.RecognitionConfig.AudioEncoding.ENCODING_UNSPECIFIED,
            ".m4a": speech.RecognitionConfig.AudioEncoding.ENCODING_UNSPECIFIED,
            ".ogg": speech.RecognitionConfig.AudioEncoding.OGG_OPUS,
        }
        
        return encoding_map.get(extension, speech.RecognitionConfig.AudioEncoding.ENCODING_UNSPECIFIED)
